src/framework/components/element/text-element.js

```diff
@@ -222,34 +229,120 @@ Object.assign(pc, function () {
+                var dict = {
+                    red: "#ff0000",
+                    green: "#00ff00",
+                    blue: "#0000ff",
+                    white: "#ffffff",
+                    black: "#000000",
+                    gray: "#808080"
+                };
+                var paletteMap = { };
+
+                // store fallback color in the palette
+                this._colorPalette = [
+                    Math.round(this._color.r * 255),
+                    Math.round(this._color.g * 255),
+                    Math.round(this._color.b * 255)
+                ];
+                this._symbolColors = [];
+                paletteMap[this._color.toString(false).toLowerCase()] = 0;
+
+                for (i = 0; i < this._symbols.length; ++i) {
+                    var tag = tags[i];
+                    var color = 0;
+
+                    // get markup coloring
+                    if (tag.color && tag.color.value) {
+                        var c = tag.color.value;
+
+                        // resolve color dictionary names
+                        if (dict.hasOwnProperty(c)) {
+                            c = dict[c];
+                        }
 
-            var textLength = symbols.length;
+                        // convert hex color
+                        if (c.length === 7 && c[0] === "#") {
+                            var hex = c.substring(1).toLowerCase();
 
-            // handle null string
-            if (textLength === 0) {
-                textLength = 1;
-                text = " ";
-                symbols = [text];
+                            if (paletteMap.hasOwnProperty(hex)) {
+                                // color is already in the palette
+                                color = paletteMap[hex];
+                            } else {
+                                if (/^([0-9a-f]{2}){3}$/.test(hex)) {
+                                    // new color
+                                    color = this._colorPalette.length / 3;
+                                    paletteMap[hex] = color;
+                                    this._colorPalette.push(parseInt(hex.substring(0, 2), 16));
+                                    this._colorPalette.push(parseInt(hex.substring(2, 4), 16));
+                                    this._colorPalette.push(parseInt(hex.substring(4, 6), 16));
+                                }
+                            }
+                        }
+                    }
+
+                    this._symbolColors.push(color);
+                }
+            } else {
+                // no tags, therefore no per-symbol colors
+                this._colorPalette = [];
+                this._symbolColors = null;
             }
 
             var charactersPerTexture = {};
 
             var char, info, map;
-            for (i = 0; i < textLength; i++) {
-                char = symbols[i];
+            for (i = 0; i < this._symbols.length; i++) {
```

Comment:
Cache `this._symbols.length` because it will be evaluated on every iteration

Reply:
done